src/FetchInterceptor.js

```diff
@@ -0,0 +1,327 @@
+
+/**
+ * Provides a default implementation for intercepting fetch requests. It will try to resolve
+ * unauthorized responses by renewing the access token and repeating the initial request.
+ */
+export default class FetchInterceptor {
+  constructor(fetch) {
+    // stores reference to vanilla fetch method
+    this.fetch = fetch;
+
+    this.config = {
+      fetchRetryCount: 1,
+      createAccessTokenRequest: null,
+      shouldIntercept: null,
+      shouldInvalidateAccessToken: null,
+      parseAccessToken: null,
+      authorizeRequest: null,
+      onAccessTokenChange: null,
+      onResponse: null,
+    };
+
+    this.intercept = this.intercept.bind(this);
+
+    this.resolveIntercept = this.resolveIntercept.bind(this);
+    this.fetchWithRetry = this.fetchWithRetry.bind(this);
+    this.isConfigValid = this.isConfigValid.bind(this);
+    this.createRequestUnit = this.createRequestUnit.bind(this);
+    this.shouldIntercept = this.shouldIntercept.bind(this);
+    this.authorizeRequest = this.authorizeRequest.bind(this);
+    this.shouldFetch = this.shouldFetch.bind(this);
+    this.fetchRequest = this.fetchRequest.bind(this);
+    this.shouldInvalidateAccessToken = this.shouldInvalidateAccessToken.bind(this);
+    this.invalidateAccessToken = this.invalidateAccessToken.bind(this);
+    this.handleUnauthorizedRequest = this.handleUnauthorizedRequest.bind(this);
+    this.handleResponse = this.handleResponse.bind(this);
+  }
+
+  /**
+   * Configures fetch interceptor with given config object. All required properties can optionally
+   * return a promise which will be resolved by fetch interceptor automatically.
+   *
+   * @param config
+   *
+   * (Required) Prepare fetch request for renewing new access token
+   *   createAccessTokenRequest: (refreshToken) => request,
+   *
+   * (Required) Parses access token from access token response
+   *   parseAccessToken: (response) => accessToken,
+   *
+   * (Required) Defines whether interceptor will intercept this request or just let it pass through
+   *   shouldIntercept: (request) => boolean,
+   *
+   * (Required) Defines whether access token will be invalidated after this response
+   *   shouldInvalidateAccessToken: (response) => boolean,
+   *
+   * (Required) Adds authorization for intercepted requests
+   *   authorizeRequest: (request) => authorizedRequest,
+   *
+   * Number of retries after initial request was unauthorized
+   *   fetchRetryCount: 1,
+   *
+   * Event invoked when access token has changed
+   *   onAccessTokenChange: null,
+   *
+   * Event invoked when response is resolved
+   *   onResponse: null,
+   *
+   */
+  configure(config) {
+    this.config = { ...this.config, ...config };
+
+    if (!this.isConfigValid(this.config)) {
+      throw new Error(ERROR_INVALID_CONFIG);
+    }
+
+    this.accessTokenProvider = new AccessTokenProvider(this.fetch, this.config);
+  }
+
+  /**
+   * Authorizes fetch interceptor with given renew token
```

Comment:
Oh... that's find/replace gone wild... Nice catch :) Fixed in class description also...